modules/promotion/src/Element/CouponRedemptionForm.php

```diff
@@ -166,26 +214,39 @@ public static function validateForm(array &$element, FormStateInterface $form_st
    *   The current state of the form.
    */
   public static function submitForm(array &$element, FormStateInterface $form_state) {
-    $coupon_parents = array_merge($element['#parents'], ['code']);
-    $coupon_code = $form_state->getValue($coupon_parents);
-    /** @var \Drupal\commerce_order\Entity\OrderInterface $order */
-    $order = $form_state->get('order');
+    $triggering_element = $form_state->getTriggeringElement();
     $entity_type_manager = \Drupal::entityTypeManager();
-    $order_type_storage = $entity_type_manager->getStorage('commerce_order_type');
-    /** @var \Drupal\commerce_promotion\PromotionStorageInterface $promotion_storage */
-    $promotion_storage = $entity_type_manager->getStorage('commerce_promotion');
-    /** @var \Drupal\commerce_promotion\CouponStorageInterface $coupon_storage */
-    $coupon_storage = $entity_type_manager->getStorage('commerce_promotion_coupon');
-
-    $coupon = $coupon_storage->loadByCode($coupon_code);
-    /** @var \Drupal\commerce_order\Entity\OrderTypeInterface $order_type */
-    $order_type = $order_type_storage->load($order->bundle());
-    $promotion = $promotion_storage->loadByCoupon($order_type, $order->getStore(), $coupon);
+    $order_storage = $entity_type_manager->getStorage('commerce_order');
+    /** @var \Drupal\commerce_order\Entity\OrderInterface $order */
+    $order = $order_storage->load($element['#order_id']);
 
-    if (self::couponApplies($order, $promotion, $coupon)) {
-      $order->get('coupons')->appendItem($coupon);
+    if ($triggering_element['#name'] == 'remove_coupon') {
+      $order->get('coupons')->setValue([]);
       $order->save();
-      drupal_set_message(t('Coupon applied'));
+    }
+    else {
+      $coupon_parents = array_merge($element['#parents'], ['code']);
+      $coupon_code = $form_state->getValue($coupon_parents);
+      if ($coupon_code) {
+        $order_type_storage = $entity_type_manager->getStorage('commerce_order_type');
+        /** @var \Drupal\commerce_promotion\PromotionStorageInterface $promotion_storage */
+        $promotion_storage = $entity_type_manager->getStorage('commerce_promotion');
+        /** @var \Drupal\commerce_promotion\CouponStorageInterface $coupon_storage */
+        $coupon_storage = $entity_type_manager->getStorage('commerce_promotion_coupon');
+
+        $coupon = $coupon_storage->loadByCode($coupon_code);
+        if ($coupon) {
+          /** @var \Drupal\commerce_order\Entity\OrderTypeInterface $order_type */
+          $order_type = $order_type_storage->load($order->bundle());
+          $promotion = $promotion_storage->loadByCoupon($order_type, $order->getStore(), $coupon);
+
+          if (self::couponApplies($order, $promotion, $coupon)) {
+            $order->get('coupons')->appendItem($coupon);
+            $order->save();
```

Comment:
Order should only be saved in rebuildForm (rename it) when element button clicked.